### PR TITLE
Chainable api support

### PR DESCRIPTION
**Pull request by fralonra**
The original issue is https://github.com/fastify/light-my-request/issues/50.

And I'm glad to update the document and typings if this feature is accepted.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -90,6 +98,57 @@ function inject (dispatchFunc, options, callback) {
...
+  doInject(this.dispatch, this.option, callback)
+}
+
+Object.getOwnPropertyNames(Promise.prototype).forEach(method => {
+  if (method === 'constructor') return
+  Chain.prototype[method] = function (...args) {
+    return doInject(this.dispatch, this.option)[method](...args)
+  }
+})
+
```
Comment (fralonra) [inline index.js]: Provides backward compatibility for promise.
Tell me if there is a better way!

File: index.js [hunk 2]
```diff
@@ -90,6 +98,57 @@ function inject (dispatchFunc, options, callback) {
...
+Chain.prototype.end = function (callback) {
+  doInject(this.dispatch, this.option, callback)
+}
+
+Object.getOwnPropertyNames(Promise.prototype).forEach(method => {
+  if (method === 'constructor') return
+  Chain.prototype[method] = function (...args) {
+    return doInject(this.dispatch, this.option)[method](...args)
+  }
+})
```
Comment (fralonra) [inline index.js]: Provides backward compatibility for promise.
Tell me if there is a better way!

File: index.js (hunk 1 above)
Comment (fralonra) [inline index.js]: Provides backward compatibility for promise.
Tell me if there is a better way!

File: index.js (hunk 2 above)
Comment (fralonra) [inline index.js]: Provides backward compatibility for promise.
Tell me if there is a better way!

Comment (mcollina) [reply index.js]: I think this is ok. I would just have added a `then` method, but this is better.

The implementation has a side effect, as invoking multiple times `.then()` will issue multiple calls vs waiting for the same one. You'll have to get hold of a promise and the keep calling the methods on that one instead, i.e. the same `Chain` object wraps only one promise.

File: test/test.js [hunk 3]
```diff
@@ -840,7 +851,17 @@ test('should throw on unknown HTTP method', (t) => {
   t.plan(1)
   const dispatch = function (req, res) { }
 
-  t.throws(() => inject(dispatch, { method: 'UNKNOWN_METHOD', url: 'http://example.com:8080/hello' }), Error)
+  t.throws(() => inject(dispatch, { method: 'UNKNOWN_METHOD', url: 'http://example.com:8080/hello' }, (err, res) => {
+    t.ok(err)
+  }), Error)
+})
```
Comment (mcollina) [inline test/test.js]: Why this change?

Comment (fralonra) [reply test/test.js]: Because with the new api, if there is no callback function, it would be treated as using chainable apis (as shown in the next test).

Comment (fralonra) [reply test/test.js]: I believe that it would be better to add `.json()` handler of `res` in another PR, for it has no concern with chainable apis. Do you agree?

File: index.js (hunk 2 above)
Comment (fralonra) [reply index.js]: Got it. But I'm wondered is there any cases should we write codes like that?

However, if different `then` returns different Promise, we can test different cases using one `Chain` Instance, just like:
```js
const chain = inject(dispatch)
chain.get('/').then()
chain.post('/').then()
```
I think the above case might be more common.

Comment (mcollina) [reply index.js]: I think we should throw in the above case, you can’t set the method twice

**Comment by mcollina** [comment]
@fralonra any update on this?

**Comment by mcollina** [comment]
Would you mind adding a test for the multiple promise fix? Also, this would need a rebase.

**Comment by fralonra** [comment]
I tried to write test like below:
```js
test('chainable api: multiple call of then should returns the same promise', (t) => {
  t.plan(1)

  function dispatch (req, res) {
    res.writeHead(200, { 'Content-Type': 'text/plain' })
    res.end('hello')
  }

  const chain = inject(dispatch).get('/')
  chain.then(res => {
    chain.then(rep => {
      t.equal(res, rep)
    })
  })
})
```
But it finally outpus the error `not ok test unfinished`.

**Comment by fralonra** [comment]
There is another consideration. Currently the usage of `end()` may conflict with `then()`.

`end()` will invoke the injection as many times as this method being called, so that we can use only one `Chain` to test different cases. That's exactly what I wanted at the begining:
```js
function dispatch (req, res) {
  // has different logic for route '/one-route' and '/another-route'
}

const chain = inject(dispatch)
chain.get('/one-route').end(() => {
  // test code
})
chain.get('/another-route').end(() => {
  // test code
})
```

Obviously it conflicts with the design that multiple `then()` returns the same promise.

**Comment by mcollina** [comment]
I think we should throw if end() is called multiple times.

File: index.js [hunk 4]
```diff
@@ -90,6 +98,59 @@ function inject (dispatchFunc, options, callback) {
...
+  doInject(this.dispatch, this.option, callback)
+}
+
+Object.getOwnPropertyNames(Promise.prototype).forEach(method => {
+  if (method === 'constructor') return
+  Chain.prototype[method] = function (...args) {
+    if (!this._promise) {
+      this._promise = doInject(this.dispatch, this.option)[method](...args)
+    }
+    return this._promise
```
Comment (mcollina) [inline index.js]: you need to call the method on promise, passing the args to it anyway.

Comment (fralonra) [reply index.js]: Sorry for forgetting changing the code.

**Comment by fralonra** [comment]
> I think we should throw if end() is called multiple times.

Should we just throw the error, or use the callback to wrap it?

**Comment by mcollina** [comment]
I think we should just throw.

File: index.js [hunk 5]
```diff
@@ -138,16 +139,21 @@ chainMethods.forEach(method => {
 })
 
 Chain.prototype.end = function (callback) {
+  if (this._hasInvoked === true) {
+    throw new Error('The dispatch function has already been invoked')
+  }
   doInject(this.dispatch, this.option, callback)
+  this._hasInvoked = true
```
Comment (mcollina) [inline index.js]: Can you make the chain return itself here? So that we can:

```js
await chain.end()
```

File: index.js [hunk 6]
```diff
@@ -138,16 +139,21 @@ chainMethods.forEach(method => {
 })
 
 Chain.prototype.end = function (callback) {
+  if (this._hasInvoked === true) {
+    throw new Error('The dispatch function has already been invoked')
+  }
   doInject(this.dispatch, this.option, callback)
```
Comment (mcollina) [inline index.js]: we should probably store the returned promise here.

File: index.js [hunk 7]
```diff
@@ -139,17 +139,25 @@ chainMethods.forEach(method => {
...
+  return this
 }
 
 Object.getOwnPropertyNames(Promise.prototype).forEach(method => {
   if (method === 'constructor') return
   Chain.prototype[method] = function (...args) {
     if (!this._promise) {
+      if (this._hasInvoked === true) {
+        throw new Error('The dispatch function has already been invoked')
+      }
```
Comment (mcollina) [inline index.js]: This is not correct. The difference between a promise and end is that one would throw, and one does not, because promises keep the resolved value.

Comment (fralonra) [reply index.js]: Since `end()` would throw if called multiple times, and multiple `then()` returns only one promise, what should we do if one calls `then()` after invoking `end()`? Should we just reject here?

Or these two methods do not effect each other, one can call `then()` after `end()` as normal, v.v.

Comment (mcollina) [reply index.js]: If inside end() you store the promise inject() returns, you can return it here.

Comment (fralonra) [reply index.js]: That's what exactly these method do currently. If you pass no callback function to `end`, it would store the promise, and return the promise in the next `then` call.

Comment (fralonra) [reply index.js]: That's what exactly these method do currently. If you pass no callback function to end, it would store the promise, and return the promise in the next then call.

Is there anything I misunderstand?

Comment (mcollina) [reply index.js]: .end() needs to set the promise as well.

Comment (fralonra) [reply index.js]: Even if there is already a callback?

Comment (mcollina) [reply index.js]: That’s an interesting case, as mixing promises and callbacks would not go well in application. I think we should just set the promise if end() is called without a callback. I would also make the error more explicit to cover the above case.

Comment (fralonra) [reply index.js]: The above case has already been handled. The application would store the promise if `end()` is called without a callback, and then we can access the promise through `then()`.

**Comment by Eomm** [comment]
@fralonra Could you check the conflicts plz?

File: README.md [hunk 8]
```diff
@@ -147,6 +162,15 @@ Note: You can also pass a string in place of the `options` object as a shorthand
 
 Checks if given object `obj` is a *light-my-request* `Request` object.
 
+#### Method chaining
+
+There are following methods you can used as chaining:
+- `delete`, `get`, `head`, `options`, `patch`, `post`, `put`, `trace`. They will set the HTTP request method and also the request URL.
+- `body`, `headers`, `payload`, `query`. They can be used to set the request options object.
+And finally you need to call `end`. It has the signature `function (callback)`.
```
Comment (mcollina) [inline README.md]: end() can be invoked without a callback

Comment (fralonra) [reply README.md]: That's right. But I think that's not the way we should recommend users to do, so I would not mention it in the document.

If invoking `end()` without a callback, users should access the promise through methods like `then()` later on:
```js
inject(dispatch)
  .get('/')
  .end()
  .then(res => {})
```

It acts the same as:
```js
inject(dispatch)
  .get('/')
  .then(res => {})
```
So I think there is no need to call `end()`. We just provide a way that makes the application more flexible.

File: index.js (hunk 5 above)
Comment (delvedor) [reply index.js]: I agree, `end` should not return `this`, because you shouldn't be able to set more things once you send the request.
At the moment, this is valid:
```js
inject(dispatch)
  .get('/')                   // set the request method to GET, and request URL to '/'
  .headers({ foo: 'bar' })    // set the request headers
  .payload('request payload') // set the request payload
  .body('request body')       // alias for payload
  .query({ foo: 'bar' })      // set the query parameters
  .end((err, res) => {
    console.log(res.payload)
  })
  .query({ foo: 'bar' }
```

Comment (fralonra) [reply index.js]: If `end()` don't return `this`, then I think we should always set the callback function in `end()`, thus there is no need to store the promise in `end()`, which is requested in https://github.com/fastify/light-my-request/pull/53#discussion_r335347129

Comment (Eomm) [reply index.js]: Why we don't return `this._promise`?

In this way:
- we stop the chaining
- we can `const res = await chain.end()`

Comment (fralonra) [reply index.js]: @Eomm Great idea! I agree.

File: index.js [hunk 9]
```diff
@@ -90,6 +98,73 @@ function inject (dispatchFunc, options, callback) {
...
+  'head',
+  'options',
+  'patch',
+  'post',
+  'put',
+  'trace'
+]
+
+httpMethods.forEach(method => {
+  Chain.prototype[method] = function (url) {
```
Comment (mcollina) [inline index.js]: I think we should throw if this is called after the function is invoked.

File: index.js [hunk 10]
```diff
@@ -90,6 +98,73 @@ function inject (dispatchFunc, options, callback) {
...
+const chainMethods = [
+  'body',
+  'headers',
+  'payload',
+  'query'
+]
+
+chainMethods.forEach(method => {
+  Chain.prototype[method] = function (value) {
+    this.option[method] = value
```
Comment (mcollina) [inline index.js]: Also this should throw

File: README.md [hunk 11]
```diff
@@ -43,6 +43,21 @@ try {
 }
 ```
 
+You can also use chaining methods if you do not pass the callback function. Check [here](#method-chaining) for details.
+
+```js
+// chaining methods
+inject(dispatch)
+  .get('/')                   // set the request method to GET, and request URL to '/'
```
Comment (delvedor) [inline README.md]: I would write two examples, one with `get` and one with `post`, since this one is not correct (you can't have a body in a `GET` request).

File: test/index.test-d.ts [hunk 12]
```diff
@@ -21,6 +21,15 @@ inject(dispatch, { method: 'get', url: '/' }, (err, res) => {
   expectType<Function>(res.json)
 })
 
-expectType<Promise<Response>>(inject(dispatch, { method: 'get', url: '/' }))
+inject(dispatch)
+  .get('/')
+  .end((err, res) => {
+    expectType<Error>(err)
```
Comment (delvedor) [inline test/index.test-d.ts]: Here you can use `expectError` :)

File: test/test.js [hunk 13]
```diff
@@ -909,6 +930,271 @@ test('Should throw if both path and url are missing', (t) => {
...
+  inject(dispatch)
+    .get('/')
+    .catch(err => t.ok(err))
+})
+
+test('chainable api: multiple call of then should returns the same promise', (t) => {
+  t.plan(1)
+
+  function dispatch (req, res) {
+    res.writeHead(200, { 'Content-Type': 'text/plain' })
```
Comment (delvedor) [inline test/test.js]: Can you also add a check to be 100% sure that dispatch has been called only once?

File: test/index.test-d.ts (hunk 12 above)
Comment (fralonra) [reply test/index.test-d.ts]: I just followed L18 to keep the coding style consistent. If using `expectError` is better, I suggest to replace these all.

File: test/test.js [hunk 14]
```diff
@@ -956,18 +956,19 @@ test('chainable api: backwards compatibility for promise (catch)', (t) => {
...
 })
 
-test('chainable api: multiple call of then should returns the same promise', (t) => {
+test('chainable api: multiple call of then should return the same promise', (t) => {
   t.plan(1)
+  let id = 0
 
   function dispatch (req, res) {
-    res.writeHead(200, { 'Content-Type': 'text/plain' })
+    res.writeHead(200, { 'Content-Type': 'text/plain', 'Request-Id': id++ })
```
Comment (mcollina) [inline test/test.js]: can you add a `t.pass()` here and increase the assertion count as well?

File: README.md [hunk 15]
```diff
@@ -43,6 +43,27 @@ try {
...
+    console.log(res.payload)
+  })
+
+inject(dispatch)
+  .post('/')                  // set the request method to POST, and request URL to '/'
+  .payload('request payload') // set the request payload
+  .body('request body')       // alias for payload
+  .end((err, res) => {
+    console.log(res.payload)
+  })
```
Comment (mcollina) [inline README.md]: It would be good to add an example with await here

File: index.js [hunk 16]
```diff
@@ -90,11 +98,88 @@ function inject (dispatchFunc, options, callback) {
...
+  'patch',
+  'post',
+  'put',
+  'trace'
+]
+
+httpMethods.forEach(method => {
+  Chain.prototype[method] = function (url) {
+    if (this._hasInvoked === true || this._promise) {
+      throwIfAlreadyInvoked()
```
Comment (mcollina) [inline index.js]: This is not ok, you are missing a return statement.

File: index.js [hunk 17]
```diff
@@ -90,11 +98,88 @@ function inject (dispatchFunc, options, callback) {
...
+  'body',
+  'headers',
+  'payload',
+  'query'
+]
+
+chainMethods.forEach(method => {
+  Chain.prototype[method] = function (value) {
+    if (this._hasInvoked === true || this._promise) {
+      throwIfAlreadyInvoked()
```
Comment (mcollina) [inline index.js]: missing return

File: index.js [hunk 18]
```diff
@@ -90,11 +98,88 @@ function inject (dispatchFunc, options, callback) {
...
+    }
+    this.option[method] = value
+    return this
+  }
+})
+
+Chain.prototype.end = function (callback) {
+  if (this._hasInvoked === true || this._promise) {
+    throwIfAlreadyInvoked()
```
Comment (mcollina) [inline index.js]: missing return

File: index.js [hunk 19]
```diff
@@ -90,11 +98,88 @@ function inject (dispatchFunc, options, callback) {
...
+    return this._promise
+  }
+}
+
+Object.getOwnPropertyNames(Promise.prototype).forEach(method => {
+  if (method === 'constructor') return
+  Chain.prototype[method] = function (...args) {
+    if (!this._promise) {
+      if (this._hasInvoked === true) {
+        throwIfAlreadyInvoked()
```
Comment (mcollina) [inline index.js]: missing return

File: index.js (hunk 16 above)
Comment (fralonra) [reply index.js]: Em...I wonder is there any particular reason that we should add a return statement after `throw`.

Comment (mcollina) [reply index.js]: ok

**Comment by Eomm** [comment]
Wasn't this feature a major since:

```
const prom = fastify.inject('/hello');
```

now require:
```
await prom.end()
```
?

**Comment by mcollina** [comment]
No, `await fastify.inject(‘/hello’)` would work as expected.
